### PR TITLE
Gpu support

### PR DESCRIPTION
**Pull request by windreamer**
@mckelvin pls review

@bhack this is what i can **guess** for a GPU cluster

cf: https://github.com/douban/tfmesos/issues/1

### CODE REVIEW

**Comment by windreamer** [comment]
Note: Mesos should be configured as below:
https://github.com/NVIDIA/nvidia-docker/issues/60#issuecomment-200485411

**Comment by mckelvin** [comment]
👍 

@bhack Would you please try this? Since we don't have GPU cluster yet.

**Comment by bhack** [comment]
Yes sure. /cc @mtamburrano @lenlen

**Comment by xunzhang** [comment]
🍺

**Comment by windreamer** [comment]
a major issue i found is TF always use the **first a few** GPU cores, so when u get several TF compute at the same time, they will all run in the same cores, with other cores in idle.

i can not find a way to specify GPU UUIDs of a TF programs to run for now.

cf: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/common_runtime/gpu/gpu_device.cc#L553

**Comment by wsxiaozhang** [comment]
I'm curious which mesos release you are working on for GPU resource. Seems mesos hasn't finished [https://issues.apache.org/jira/browse/MESOS-4424](url)

**Comment by windreamer** [comment]
@wsxiaozhang we can add GPU cores as custom resources for Mesos before MESOS-4424 ships.
of coz it is not perfect, no isolator for now, but barely usable i think.

**Comment by bhack** [comment]
@windreamer With manual  placement GPU can be addresses with id i.e. in the form of /gpu:0. See https://www.tensorflow.org/versions/r0.8/how_tos/using_gpu/index.html

**Comment by windreamer** [comment]
@bhack yes, but imagine u have 2 TF programs running, ur machine have 4 GPUS. Ideally we allocate 2 GPUs to the first program as "gpu:0" and "gpu:1" and the other 2 GPUs to the other program also as "gpu:0" and "gpu:1".

But in fact, the two program will both use GPU0 and GPU1, and GPU2, GPU3 will be idle.

**Comment by bhack** [comment]
But the id of the GPU free resources offered by mesos don't change? Or Is it ignored by the sequential internal cycle of TF?

**Comment by windreamer** [comment]
yes, u can see the TF code i mentioned. no gpu-id is aware there unfortunately. maybe a docker isolator can do the magic to reassign ids to allocated gpus? but i'm not optimistic.

**Comment by bhack** [comment]
Do you mean i.e. that magically what it is presented as GPU 0 in the container it is GPU 3 in the host etc..?

**Comment by bhack** [comment]
Probably @3xx0 could give us some hint on this for nvidia-docker.

**Comment by 3XX0** [comment]
You can use `nvidia-docker` instead of `docker` in the mesos-slave configuration.
Regarding Mesos resources, either use the GPGPU support in Mesos upstream

http://c99.millennium.berkeley.edu/documentation/latest/gpu-support/
https://issues.apache.org/jira/browse/MESOS-4424

or if you need more information about which GPUs to pick, use the custom resource advertised by `nvidia-docker-plugin`:

https://github.com/NVIDIA/nvidia-docker/issues/60

**Comment by bhack** [comment]
In the link i read that "Filesystem isolation must be disabled. This means No Docker Support at the moment. Only the mesos containerizer can currently be used."

**Comment by 3XX0** [comment]
No Docker support through the Mesos containerizer.
If you use the Docker containerizer or the Mesos containerizer without filesystem isolation you're fine.

**Comment by windreamer** [comment]
@3XX0 Does `nvidia-docker` re-map GPUs inside containers? For example, `/dev/gpu3` in host machine will be `/dev/gpu0` inside container?

**Comment by windreamer** [comment]
According this anwser: http://stackoverflow.com/a/26568684
Seems we can do the re-mapping in `tfmesos`
But we should use gpu id instead of gpu uuid for resources.

**Comment by windreamer** [comment]
@bhack some idea updated here. Note: mesos gpu resources should be defined as below:
```
--resources=gpus:{0, 1, 2, ...}
```

**Comment by bhack** [comment]
As @3XX0 confirmed seems to me that with current limitation GPU cannot work in the case we set DOCKER_IMAGE env variable. Is this right?

**Comment by windreamer** [comment]
@bhack seems we can use Docker containerizer with nvidia-docker

**Comment by bhack** [comment]
"If you use the Docker containerizer with nvidia-docker or the Mesos containerizer without Docker and without filesystem isolation you're fine." This seems to me that on slave we cannot run Task inside docker or nvidia-docker containers. Do you have another interpretation?

**Comment by windreamer** [comment]
My interpretation is 
> If you use the Docker containerizer with nvidia-docker  you're fine.
or
> If you use the Mesos containerizer without Docker and without filesystem isolation you're fine.

so I think we can go with the first option.

**Comment by bhack** [comment]
If your interpretation is correct how can we let the mesos slave to use nvidia-docker? I think that slave simply invoke simply docker command when your slave run with --containerizers=docker,mesos

**Comment by windreamer** [comment]
@bhack `nvidia-docker` seems a plugin for docker provided by Nvidia for CUDA. 
cf: https://github.com/NVIDIA/nvidia-docker/wiki/Installation

but I have no experience to build up a Mesos cluster with CUDA support, I am not sure about it.

**Comment by windreamer** [comment]
cf: https://github.com/NVIDIA/nvidia-docker/issues/8

**Comment by bhack** [comment]
Yes the plugin Rest API can prepare the GPU arguments fom docker. Generally on the GPU node without mesos you can run nvidia-docker command line instead of docker. But here, with tfmesos we are talking about a container that need to be runned on slave by mesos. The only solution is to [override CommandInfo](http://mesos.apache.org/documentation/latest/docker-containerizer/) to use nvidia-docker. 
But we need to carefully handle this only for GPU targets because on CPU only node we need leave it default.

**Comment by 3XX0** [comment]
Sorry I should have been clearer, @windreamer is correct you can go with the first option.
You can use the following and you'll have everything you need. Note that the `nvidia-docker-plugin` daemon must be running on the host.
```
mesos-slave --docker=/usr/bin/nvidia-docker --containerizers=docker,mesos `curl -s localhost:3476/mesos/cli`
```

Basically the framework should pick one of the UUID in the resource named `gpu`, lookup its index in the custom attribute `gpu` and sets the environment variable `NV_GPU` with the GPU index in the executor info. When containers are launched GPU are renumbered inside the container.

**Comment by windreamer** [comment]
@3XX0 This is quite helpful, I will try this way, thx.

**Comment by bhack** [comment]
@3XX0 Thank you it is much cleaner now. Some of this info could be helpful to be available in the official docs of mesos/Nvidia docker.

**Comment by windreamer** [comment]
@bhack this is a bit tricky. and we are going on vacation next week here in China. I will try to update this PR later.

**Comment by lenlen** [comment]
@3XX0 I try to run the slave with --docker=/usr/bin/nvidia-docker, but I got this error:

```
Failed to update resources for container e4382098-dfd4-4b49-b138-d7832838b173 of executor '1' running task 1 on status update for terminal task, destroying container: Failed to '/usr/bin/nvidia-docker -H unix:///var/run/docker.sock inspect mesos-8191b370-f511-427f-ad7a-f0b83f451748-S0.e4382098-dfd4-4b49-b138-d7832838b173': exit status = exited with status 1 stderr = Error: No such image or container: mesos-8191b370-f511-427f-ad7a-f0b83f451748-S0.e4382098-dfd4-4b49-b138-d7832838b173
```

if I use the slave with  --docker=/usr/bin/docker the run is ok

**Comment by 3XX0** [comment]
Oh right I forgot, for some reason the `PATH` is unset inside the executor, you also need to add:
`--executor_environment_variables='{"NV_DOCKER": "/usr/bin/docker"}'`

**Comment by bhack** [comment]
@3XX0 Thank you, I've not seen that PATH was missing. @windreamer Ok, ping us where you are back again

**Comment by windreamer** [comment]
@bhack i'm back now. how is everything now?

**Comment by lenlen** [comment]
@windreamer the framework is registered by the scheduler but the socket is always not readable, so in https://github.com/windreamer/tfmesos/blob/gpu/tfmesos/scheduler.py#L270 the readable function returns always False.

**Comment by windreamer** [comment]
@lenlen when TFMesos task is successfully started on Mesos, it should connect back to scheduler via hostname (cf: https://github.com/windreamer/tfmesos/blob/gpu/tfmesos/server.py#L24 ). So you'd better provide some more information to help me identify the problem:

1. can u connect to the machine via hostname on the slave server of Mesos?
1. is any task the scheduler submitted to Mesos in RUNNING status?
1. how is the logs looks like on Mesos master & Mesos slave & Mesos Executor?

**Comment by lenlen** [comment]
1 - Yes I'm connecting to the machine via hostname
2 - The framework is registered, I see it in the web gui but the active tasks are 0
3 - Logs:

Master:
```
I0523 11:40:12.428740 20741 master.cpp:2231] Received SUBSCRIBE call for framework '[tensorflow] /tmp/demo.py NN-computer.crrlab' at scheduler@192.168.7.1:367
51
I0523 11:40:12.428797 20741 master.cpp:2302] Subscribing framework [tensorflow] /tmp/demo.py NN-computer.crrlab with checkpointing disabled and capabilities [  ]
I0523 11:40:12.428992 20736 hierarchical.cpp:265] Added framework 55508a8a-604a-4c37-bd7b-ea5b2465a5a7-0012
I0523 11:40:12.429215 20736 master.cpp:5324] Sending 2 offers to framework 55508a8a-604a-4c37-bd7b-ea5b2465a5a7-0012 ([tensorflow] /tmp/demo.py NN-computer.crrlab) at scheduler@192.168.7.1:36751
I0523 11:40:21.150558 20745 http.cpp:312] HTTP GET for /master/state from 192.168.7.3:58546 with User-Agent='Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:45.0) Gecko/20100101 Firefox/45.0'
...
I0523 11:42:13.817292 20744 master.cpp:1212] Framework 55508a8a-604a-4c37-bd7b-ea5b2465a5a7-0012 ([tensorflow] /tmp/demo.py NN-computer.crrlab) at scheduler@192.168.7.1:36751 disconnected
I0523 11:42:13.817373 20744 master.cpp:2527] Disconnecting framework 55508a8a-604a-4c37-bd7b-ea5b2465a5a7-0012 ([tensorflow] /tmp/demo.py NN-computer.crrlab) at scheduler@192.168.7.1:36751
I0523 11:42:13.817387 20744 master.cpp:2551] Deactivating framework 55508a8a-604a-4c37-bd7b-ea5b2465a5a7-0012 ([tensorflow] /tmp/demo.py NN-computer.crrlab) at scheduler@192.168.7.1:36751
I0523 11:42:13.817438 20740 hierarchical.cpp:375] Deactivated framework 55508a8a-604a-4c37-bd7b-ea5b2465a5a7-0012
W0523 11:42:13.817487 20744 master.hpp:1764] Master attempted to send message to disconnected framework 55508a8a-604a-4c37-bd7b-ea5b2465a5a7-0012 ([tensorflow] /tmp/demo.py NN-computer.crrlab) at scheduler@192.168.7.1:36751
W0523 11:42:13.817661 20744 master.hpp:1764] Master attempted to send message to disconnected framework 55508a8a-604a-4c37-bd7b-ea5b2465a5a7-0012 ([tensorflow] /tmp/demo.py NN-computer.crrlab) at scheduler@192.168.7.1:36751
I0523 11:42:13.817749 20744 master.cpp:1236] Giving framework 55508a8a-604a-4c37-bd7b-ea5b2465a5a7-0012 ([tensorflow] /tmp/demo.py NN-computer.crrlab) at scheduler@192.168.7.1:36751 100secs to failover
```

Slaves:
```
I0523 11:43:53.820289 20810 slave.cpp:2079] Asked to shut down framework 55508a8a-604a-4c37-bd7b-ea5b2465a5a7-0012 by master@192.168.7.1:5050
W0523 11:43:53.820646 20810 slave.cpp:2094] Cannot shut down unknown framework 55508a8a-604a-4c37-bd7b-ea5b2465a5a7-0012
```

**Comment by windreamer** [comment]
@lenlen 

yes, it seems framework does register and 2 offers have been send to it
```
I0523 11:40:12.429215 20736 master.cpp:5324] Sending 2 offers to framework 55508a8a-604a-4c37-bd7b-ea5b2465a5a7-0012 ([tensorflow] /tmp/demo.py NN-computer.crrlab) at scheduler@192.168.7.1:36751
```

and maybe i need to see more logs of `tfmesos`, maybe you can enable debug log and i can get more information from it.

**Comment by lenlen** [comment]
Reading the log, seems there is a problem with self.gpus of TFMesosScheduler

```
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/pymesos/process.py", line 69, in run_jobs
    func(*args, **kw)
  File "/usr/local/lib/python2.7/dist-packages/pymesos/process.py", line 142, in handle
    f(*args)
  File "/usr/local/lib/python2.7/dist-packages/pymesos/scheduler.py", line 110, in onResourceOffersMessage
    self.sched.resourceOffers(self, list(offers))
  File "/usr/local/lib/python2.7/dist-packages/tfmesos/scheduler.py", line 212, in resourceOffers
    gpus = int(math.ceil(self.gpus))
AttributeError: 'TFMesosScheduler' object has no attribute 'gpus'
```

It works on cpu if I change the line with

```
gpus = int(math.ceil(len(offered_gpus)))
```

It is correct?
But with this fix I have another error:

```
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/pymesos/process.py", line 69, in run_jobs
    func(*args, **kw)
  File "/usr/local/lib/python2.7/dist-packages/pymesos/process.py", line 142, in handle
    f(*args)
  File "/usr/local/lib/python2.7/dist-packages/pymesos/scheduler.py", line 110, in onResourceOffersMessage
    self.sched.resourceOffers(self, list(offers))
  File "/usr/local/lib/python2.7/dist-packages/tfmesos/scheduler.py", line 218, in resourceOffers
    offer, self.addr, gpu_ids=gpu_ids))
  File "/usr/local/lib/python2.7/dist-packages/tfmesos/scheduler.py", line 102, in to_task_info
    p.value = '/dev/nvidia%d:/dev/nvidia%d' % (gpu_id, id)
TypeError: %d format: a number is required, not unicode

```

**Comment by windreamer** [comment]
@lenlen sorry my bad. i have fix this typo

and it seems you are using this setting for mesos
```
mesos-slave --docker=/usr/bin/nvidia-docker --containerizers=docker,mesos `curl -s localhost:3476/mesos/cli`
```

but this PR is intended for a setting like this:
```
mesos-slave --resources=gpus:{0, 1, 2, ...}
```

so, would you please tell me what is the output of `curl -s localhost:3476/mesos/cli` ? and i think i can adjust the code according to the output.

**Comment by lenlen** [comment]
Yes, the output is:

````
--attributes=gpus:eJyckk9v00AQxe98imjOdrp_vPbat8ShUQRBEcSoEuKw8Y6J1a23stMEFPm7MwuiJQgKwifPjt_bN_PzGd5jP7S-g-IMi749Yg8FSCWmWQ4RlNViRnU2VTBGsMBjW-MAxYczrL1FR60lXvu-xslyezPZrrazN5Mb0lXVahGamyrOeZYgJjbmJtOxzFQe650Ucd4kxhrkaaolKTbmsCfFlcXjVXdsbWt4OPWnEEgoRlk21axp2q49fIGC6k25CqHnD8O3uxg9BVMFY1NGyvnsLQ_ClF5NZ0-tDf5cZSmjSUrn69shyEvfh4m44jqCNd75ntylYoq-2vp77_wnOugenIvg2ty1jipYm88ndI6umfV1iK2mImzru5dkmXjyOsPLsoSiMW7ACJbO7wztjQuhkwje7U2PFoo8DepuOJjuAEVKndeiNPUef7j9NIOUaaLZOEbPQZg7U99egsgablDbXYzGylg3qY3VzuaxzUWtBSaclvNbEOx_QOSXILjQ_whCMHEBgj0H4hXeO8r1yEHSj_rIQWjN_sYh5Yl8wpDoP2LgSqa_YqBBx_Hj-OJrAAAA__8wme4U --resources=gpus:{GPU-9174ee4d-1a78-3759-8b32-9f4adae16683,GPU-7f1ae8db-ead3-8f6d-5bd9-d92c82e41157}
```

**Comment by windreamer** [comment]
@lenlen i try to support `nvidia-docker-plugin` settings in the latest commit, hope this time it works

File: tfmesos/scheduler.py [hunk 1]
```diff
@@ -96,20 +100,14 @@ def to_task_info(self, offer, master_addr, gpu_ids=[]):
...
-                    p.key = 'device'
-                    p.value = path
-
                 env = ti.command.environment.variables.add()
                 env.name = 'NV_DOCKER'
                 env.value = '/usr/bin/docker'
 
+                env = ti.command.environment.variables.add()
+                env.name = 'NV_GPU'
+                env.value = ','.join(gpu_mapping[uuid] for uuid in gpu_uuids)
```
Comment (lenlen) [inline tfmesos/scheduler.py]: set doesn't have getitem method. Transform it in dict
```
gpu_mapping = dict(gpu_mapping)
```

File: tfmesos/scheduler.py [hunk 2]
```diff
@@ -174,6 +172,23 @@ def __init__(self, task_spec, master=None, name=None, quiet=False,
...
+        for a in attributes:
+            if a.name == 'gpus':
+                attr = a.text.value
+                break
+
+        if attr is None:
+            return {}
+
+        info = json.loads(zlib.decompress(
+            base64.b64decode(attr, '-_')))
```
Comment (lenlen) [inline tfmesos/scheduler.py]: there is an error for unicode string, add str to attr variable
```
 info = json.loads(zlib.decompress(
            base64.b64decode(str(attr), '-_')))
```

**Comment by lenlen** [comment]
I have commented your last commit for two errors. Then there is a problem with libcudnn.so because the library is not in the path LD_LIBRARY_PATH. It is needed add "/usr/lib/x86_64-linux-gnu" in LD_LIBRARY_PATH and create a symlink from libcudnn.so.4 from libcudnn.so.
With these changes a simple example works if I define one gpu:
```
jobs_def = [
    {
        "name": "ps",
        "num": 2
    },
    {
        "name": "worker",
        "num": 2,
        "gpus": 1
    },
  ]
```
But using 2 gpus the run fails and from the log I find:
```
Sending 2 offers to framework ...
Processing DECLINE call for offers...
```

**Comment by 3XX0** [comment]
@windreamer I just added support for UUID in `NV_GPU`, it should simplify the index lookup: https://github.com/NVIDIA/nvidia-docker/commit/1dff2bb5088c2519b1db269b35cd7f9bec60f64c

**Comment by windreamer** [comment]
@lenlen thanks for the feed back, I've fix the errors you mentioned and commit to this PR, please review my fix and try again if it is OK to you.

`libcudnn.so` problem is a bit weird to me.  I will try to investigate this problem and fix in dockerfile

further more can you provide more information abort how the task is failed when more gpus are allocated ?

**Comment by windreamer** [comment]
@3XX0 bravo! this is really convenient. thank you for this work.

but at present I may temporarily keep the code this way, I will simpfy the whole process when `nvidia-docker` v1.0 releases ASAP

**Comment by vitan** [comment]
@3XX0  

```
You can use nvidia-docker instead of docker in the mesos-slave configuration.
```

Here I have a concern to the above given solution. The above configuration will make all of containers deployed by mesos use the `nvidia-docker` instead `docker`, isn't it? And I don't think it is a general solution.

**Comment by 3XX0** [comment]
Yes and? It is a general solution until we upstream all the work in Mesos itself (which might take a while).
Right now in Mesos, we only support GPUs as a scalar with the Mesos containerizer and without filesystem isolation (see [presentation](https://www.youtube.com/watch?v=giJ4GXFoeuA&index=46&list=PLGeM09tlguZQVL7ZsfNMffX9h1rGNVqnC)).

**Comment by windreamer** [comment]
@bhack @3XX0 finally we go some GPU instance in AWS and are able to test this PR. and I made some changes:

1. upgrade tensorflow to 0.9 and upgrade other dependencies
2. thanks to `nvidia-docker=1.0.0.rc.3`, we can use UUIDs to map GPU into docker, and no longer need to decompress gpu infos
3. because when no `NV_GPU` is empty, `nvidia_docker` map all gpu into the docker, but we are expecting NO GPU should be mapped, we try to use `nvidia-docker-plugin` to get docker parameters from remote and this also enable us to setup a cluster mixing gpu nodes and cpu nodes.

so this is what we currently get, and we are planning to deliver a simple script to help setup an CPU/GPU cluster of AWS G2 ubuntu nodes later.

pls feel free to comment, any suggestion is welcome.
thx

**Comment by bhack** [comment]
I think that you can expose this news in the related  TF ticket.

**Comment by bhack** [comment]
@windreamer What about https://issues.apache.org/jira/plugins/servlet/mobile#issue/MESOS-540?

**Comment by windreamer** [comment]
@bhack sorry, anything related to this ?

**Comment by bhack** [comment]
Sorry, I've accidentally removed the last digit from URL... It is https://issues.apache.org/jira/plugins/servlet/mobile#issue/MESOS-5401

**Comment by windreamer** [comment]
@bhack OK, this is not a problem, [MESOS-5401](https://issues.apache.org/jira/plugins/servlet/mobile#issue/MESOS-5401) is to resolve isolation problem in `mesos` containerizer, not in `docker` containerizer. we are  using `nvidia-docker-plugin`, so far so good.

**Comment by bhack** [comment]
Probably it is parte of unified containers that noto require Docker daemon anymore in 1.0. See http://mesos.apache.org/blog/mesos-1-0-0-released/

**Comment by windreamer** [comment]
Hmm, `unified containers` seems just we can inject new container implementation without changing `Mesos` code. when we need `docker`, we still need `docker daemon`. further more, if we need GPU support, i will choose `nvidia docker` not the embedded GPU support from mesos.

**Comment by bhack** [comment]
Are you sure. I'havent tried yet new Mesos 1.0 but i read 
> With the unified containerizer, we can now use a  single container runtime engine, the MesosContainerizer to run containers off the most popular image formats (Docker, Appc, and soon OCI) without dependency on any external daemons.

**Comment by bhack** [comment]
And > As a bonus, we mimic the functionality of nvidia-docker so you can take your existing GPU docker containers and run them on Mesos without modification.

**Comment by windreamer** [comment]
from my [incomplete] code dig, YES, we still need docker installed.
cf: https://github.com/apache/mesos/blob/master/docs/mesos-containerizer.md#docker-runtime-isolator

**Comment by windreamer** [comment]
frankly speaking, i think everything is a MESS to me now: `docker` are integrating `swarm` to try beat `mesos` down, and `mesos` are supporting more container type and even try to parse `docker` image to control the launch process. why we can not just "God to God, Caesar to Caesar" ?

at present, i believe we should do containerization with `docker`, isolate GPUs using `nvidia-docker`, and organise containers using `mesos`. and for future? God knows!

**Comment by bhack** [comment]
@windreamer Mesos added native docker image support and nvidia-docker coverage so I think that with mesos 1.0 we don't require docker daemon anymore.
(Basically? It is a war!)[http://www.nextplatform.com/2016/07/27/mesos-reaches-milestone-adds-native-docker/]

**Comment by windreamer** [comment]
OK, i think i found the so called `native docker ` https://github.com/apache/mesos/blob/master/docs/docker-volume.md

this whole thing is too complicated, i do not think i want to give this a try.

**Comment by bhack** [comment]
There are some appreciable motivations. Read the introduction at http://mesos.apache.org/documentation/latest/container-image/

**Pull request merged**